### PR TITLE
WIP Bring back discriminated union for MathNode

### PR DESCRIPTION
**Pull request by mattvague**

### CODE REVIEW

**Comment by josdejong** [comment]
Thanks for your PR Matt, do you already want feedback on the draft?

**Comment by mattvague** [comment]
>Thanks for your PR Matt, do you already want feedback on the draft?

No problem. Yes I would like feedback, but with the caveat that I need to have another look over it myself as well.

**Comment by josdejong** [comment]
I just reviewed the PR, it looks good to me and well tested 👌, thanks!

Thoughts:

1. Do we need to describe how to implement a new Node class in TS in the documentation?
2. I'm not entirely sure if this will be a breaking change or not. Using `MathNode` is still the public API, but it's definition changed. Are there current use cases of `MathNode` that will break?

For reference: this will resolve #2810

**Comment by mattvague** [comment]
> Do we need to describe how to implement a new Node class in TS in the documentation?

Ah, yes we do -- will do that. Where do you think the most appropriate place is? Should we maybe start a new page dedicated to Typescript? 

> I'm not entirely sure if this will be a breaking change or not. Using MathNode is still the public API, but it's definition changed. Are there current use cases of MathNode that will break?

Not totally sure yet either to be honest, but I think it would pay to be cautious. I'm going to pull these changes into my own project first and see what happens, I wonder if we could round up a few other projects using mathjs + typescript extensively and see if it breaks anything for them?

File: types/index.d.ts [hunk 1]
```diff
@@ -162,36 +162,36 @@ declare namespace math {
...
   interface AccessorNodeCtor {
-    new <TObject extends MathNode = MathNode>(
+    new <TObject extends BaseNode = BaseNode>(
       object: TObject,
       index: IndexNode
     ): AccessorNode<TObject>
   }
 
-  interface ArrayNode<TItems extends MathNode[] = MathNode[]> extends MathNode {
+  interface ArrayNode<TItems extends BaseNode[] = BaseNode[]> extends BaseNode
```
Comment (ChristopherChudzicki) [inline types/index.d.ts:182]: I think there may have been an error resolving a merge conflict or something... There's a `{` missing here which is resulting in some type errors.

**Comment by ChristopherChudzicki** [comment]
Ack, sorry I missed the pings. I'm having a little trouble testing this out because I think there was an error resolving a merge conflict, and now the `index.d.ts` file is not valid. 

@josdejong  My expectation is that this
- v11.2 --> v11.3 contained a breaking change, changing `type MathNode` from a union (`AccessorNode | ArrayNode | ... `) to a looser interface.
- This essentially restores the union, but makes it extensible
- So this would be breaking relative to 11.3 (because it changes how extension is performed) but NOT breaking relative to 11.2

@mattvague Again, sorry for being slow. Had I been faster, maybe we'd have avoided the conflict all together. I should have much more time for the next 5 days or so. I'm happy to help resolve the issue, though I'm not sure the etiquette: would you want another PR into your branch? (Or happy to leave it to you.)

**Comment by josdejong** [comment]
Thanks for your clear feedback @ChristopherChudzicki . It makes sense to me to not mark this as a breaking change, but as a bug fix of the regression introduced in `v11.3`. In some way, every bug fix is a breaking change, since it alters the original behavior 😜 .

@mattvague can you still write a paragraph in the documentation page [/docs/expressions/customization.md](https://github.com/josdejong/mathjs/blob/develop/docs/expressions/customization.md)?

**Comment by josdejong** [comment]
@mattvague do you see any chance finishing up this PR? If not I can see if I can write the docs.

**Comment by ChristopherChudzicki** [comment]
**TLDR:** This only partially brings back the discriminated union, and I think the "only partially" makes the result confusing, and so maybe this isn't worth it.

---

With the goal of finishing up this PR (or starting a new one), I just took another look at this, which revealed some issues/features of the current approach that I think could be confusing.

For example, as a user consuming the library:
```ts
// Reminder: 
//  - BaseNode = generic interface
//  - MathNode = extensible collection that supports discriminated unions

const myFunctions = (node: MathNode) => {
    node.traverse(child => {
        // child has type MathNode, so discriminated unions work.
        // Cool!
    })


    if (node.type === 'ArrayNode') {
        // "items" has type BaseNode[]... Surprising!
        const items = node.items
    }
}
```

As an end user, I would expect to always get a `MathNode`, and I would be confused why sometimes I get just a `BaseNode` (whose `node.type` property is just "string", so won't support discriminated unions.)

I would **love** to have the discriminated union back: it is very convenient and makes node times much more discoverable. But IMO, the combination of 
- unpredictability of "sometimes I get BaseNode, sometimes I get MathNode"
- increased complexity from supporting an extensible discriminated union
might make this not worth it.

**Can the unpredictability be removed?**: Maybe, but I can't figure out how. Everything I've tried leads to TS complaining about circularly defined types. (E.g., changing the default type parameter of ArrayNode [here](https://github.com/josdejong/mathjs/pull/2820/files#diff-093ad82a25aee498b11febf1cdcb6546e4d223ffcb49ed69cc275ac27ce0ccceR182) to MathNode leads to TS complaining.)

**Comment by josdejong** [comment]
Thanks for digging further @ChristopherChudzicki . So, would this PR be a (little) step in the right direction you think? Or better to start from scratch?

**Comment by mattvague** [comment]
>@mattvague do you see any chance finishing up this PR? If not I can see if I can write the docs.

Hi there @josdejong, very sorry for disappearing. Unfortunately I don't see being able to work on this in the next few weeks, but if someone else can take the charge I'm happy to provide feedback if wanted. I agree the inconsistency between `BaseNode` and `MathNode` is confusing and the reason we ended up doing that (if I recall) was exactly the "circularly defined types." issue @ChristopherChudzicki mentioned, I'm not sure what to do about that at the moment. Maybe it's best to check some other OSS libraries with a similar concept and see what they do?

**Comment by josdejong** [comment]
Thanks for getting back @mattvague. No problem.

Is there anyone else who can pick this up? Help would be very welcome.

**Comment by ChristopherChudzicki** [comment]
I have been looking at this off and on. I really don't have an answer, but wanted to post the following minimal-ish reproduction of the issue: [TS playground](https://www.typescriptlang.org/play?#code/PQKhCgAIUhxB7AhgGwFxRgdQKYHIAmkyAlgNbaQCyiALgBYBy8+2AzpDfAHQ8aSQBGLpABGFbAA8a2AHatiI5Nh7CAojMIBXVtgBO7AMaIZkFgDNiMilYDuHAJ4AHNn0gAmYQGVNjx-F00psSsBrrEALaWtNhaMsTwcq4AzF4+fgGQAOayesQGkI6Iuojh2NL6kLKIipaZkIiQdMSZdHpE2ABu2MiQ8GaQrM4GxBbDNPYANJVcmVwTrvym2AbIRbX1JgCCusX2TCyQNnTwOpDE0uGGCaw0xZYxHPD1A5oiOoF99cg91PT7LtBIHxPJYDBRLOUzIgwewiuIpLJ5IpsFMWCs1jJMuhAQADPE0Vh8CF6KFgqi0RjMNiQADeAF8+ONnOS-lTIABeFmUlisADa5Hsn1+3LYAF0+HicXxoU1OtTMkhkOwABQCACUG0Iyrcaq4fD4m3w+HO8RkKGQkw4rQcznMkDM-i5-0OxG+kDNNE0xW+9kgumwZj0sjJ5x0yDMwgAYo7JCVHEp5oD6BQjN8RNDSAVEKwdIROJAcTIqVxboguvpsDjGtnnsLnTXQ-biPpAkVMppSjIaHrAcDuv1-YH-V3iCgbTEA+xtJpzb6bP5MxYrGcTAAVTxTESaQLJ-0DbDYS6PUQUVjwUp8EiRGi0U3sGznOh+7CZIr4dj5sNmAC0g6DMjBQgDHPER7kIB96ALAU+imQoAjyWdDlaGRGWtL9nyHYMKHgAwDC9dgEitCglDMGhvzoYxCHkA5PmTJZ0WKGhTR7GA+1PcJzW-YIn39RxdGYTQDCYwixGQeAbGEfVASYaQrVoIjKh2R0FWpRAbEQX0RgLOsqSrYIVxubBEEIcwwPqWFIE0OIEj4CDjm3SBoIjKSQGAcBwGJXRSQoABJHSDhpKBx1QAZblqABuILS3LHQQuVIL+FTZB0wMUg4qLFgQv8lEs3oEKbjCTFYLhLssopf4NXZAA+SAOngYh8CCyqarqhrIv4bAAEcZyVOL4F3MrWRYZrRHgeAlGMdrIBWBJsDikb6GCSKGXAUAIEBVcmnvV0emOZBCHNBTBmWEY8nHAj+mywloDczzvKdKl2HpcAmQobKOQenl+WwQVLvKx7RUi9y1uBc9sCOIMVy5AApDdDgoUzlwaAAFErk3kVh6w0eojTOHcniuljIB8mhcHYctfXkcJ42UDBbq7EloQoVHh3R4JnUkaQNHYPz-oCqKnDmyAACIWdkNnMapYWguAhnSs+7BwAZO6mYVp6grFrtWgx-4Qs1iX-iV4HgEgAARAN7kgcJ-GsR7KiVcHWj3SwYc8TUcbzVpwmPK7VpN9R8G-bQ9EMYxTCeOjWBKWngBNgAhbcVDgHIwgMdh8ASXBAnnXRM0yf1aC8A9IG2XZDdWsBXGRgAlAB5OOABlVEobEYEgb8Fd6EwG4ACTdz4VbJNFVkY01XA797EYeDpRygn6+hxIm+GR-jkW9lTYXUzS-qGvB04tuJhJMfSGipmnLOsmQWPpyFVdLjT-gAHlXSoEW5zvOWymrOdkd9ifewK-BXohWFvfPYUsgrnEPKwEKq5eTimVgzLyqsrq0iCiDRYMBVBKV0AAQgSjAYWuAwH-FwGcdgf5hyASCP6ISFpei6BXMaWhNB6Eu0bOJEwr0NhFhvEfLg0tMFuX4CQqkIVREsCVkAA)

We seem to have three goals (for this issue):
1. MathNode type is extensible, users can add more node types
2. MathNode type supports discriminated union
3. individual node types support generic parameters whose value might be `MathNode` (e.g., the `TItems` parameter of ArrayNode).

Dropping any one of these three makes the problem significantly easier—Currently, we have 1 & 3 (and previously had 2 & 3).

All of this is to say... I'm a bit stuck on this. I'll keep an eye out for potential solutions, but probably won't be actively working on it. For now, I'm just going to start narrowing types by the predicates (`isFunctionAssignmentNode`, etc, as initially discussed in https://github.com/josdejong/mathjs/issues/2810) and upgrade to the current version of MathJS.
